internal/shim-sev/src/hostmap.rs

```diff
@@ -0,0 +1,241 @@
+                }; HOST_MEM_LIST_NUM_ENTRIES],
+            },
+        }
+    }
+
+    /// Get the host virtual address for a shim physical address
+    fn get_virt_addr(&self, addr: usize) -> Option<usize> {
+        let mut free = &self.host_mem;
+        loop {
+            for i in free.ent.iter() {
+                if i.shim.end == 0 {
+                    // Found an unused map entry
+                    return None;
+                }
+                if i.shim.end > addr as usize {
+                    return Some(
+                        i.virt_start
+                            .checked_add(addr.checked_sub(i.shim.start).unwrap())
+                            .unwrap(),
+                    );
+                }
+            }
+            match free.header.next {
+                None => return None,
+                Some(ref f) => free = *f,
+            }
+        }
+    }
+}
+
+impl RWLocked<HostMap> {
+    /// Extend the number of slots in the map
+    pub fn extend_slots(&self, mem_slots: usize, allocator: &mut Heap) {
+        fn inner_extend_slots(
+            mut this: RwLockWriteGuard<HostMap>,
+            mem_slots: usize,
+            allocator: &mut Heap,
+        ) {
+            // Allocate enough pages to hold all memory slots in advance
+            // There is already one HostMemListPage present, so we can ignore the rest of the division.
+            let num_pages = mem_slots.checked_div(HOST_MEM_LIST_NUM_ENTRIES).unwrap();
+
+            if this.num_pages >= num_pages {
+                return;
+            }
+
+            let mut last_page = &mut this.host_mem as *mut HostMemListPage;
+
+            for _i in 0..num_pages {
+                unsafe {
+                    last_page = match (*last_page).header.next {
+                        None => {
+                            let new_page = {
+                                let page_res = allocator.allocate_first_fit(
+                                    Layout::from_size_align(
+                                        size_of::<HostMemListPage>(),
+                                        align_of::<HostMemListPage>(),
+                                    )
+                                    .unwrap(),
+                                );
+
+                                if page_res.is_err() {
+                                    return;
+                                }
+
+                                let page: *mut HostMemListPage = page_res.unwrap().as_ptr() as _;
+
+                                page.write_bytes(0, 1);
+                                page
+                            };
+
+                            (*last_page).header.next = Some(&mut *new_page);
+                            this.num_pages = this.num_pages.checked_add(1).unwrap();
+                            new_page
+                        }
+                        Some(ref mut p) => *p as *mut _,
+                    };
+                }
+            }
+        }
```

Comment:
I think it's harder to read a function when there's another function inside it larger than a reasonably simple lambda. I would suggest just making it a private method within the impl block instead, or just put its contents between the unsafe fences.

Reply:
ok

Reply:
Done.. Thanks!